docs/is_thread_safe/README.md

```diff
@@ -0,0 +1,132 @@
+# Is Stretch Body thread safe?
+
+The short answer: no. This tutorial builds up to a few kinds of threading experiments that prove this, and hopefully will serve as a launching point for making this library thread-safe.
+
+First, let's define what it means to be thread-safe. If we create an instance of the `Robot` class, we'd like to know if we can safely use it in Python threads. Using the `Robot`, there's two ways we interact with the underlying hardware: reading sensor data and writing motion commands. Our experiments will cover:
+
+ 1. Single threaded reading and writing
+ 2. Single writer, but multithreaded readers
+ 3. Multithreading for both reading and writing
+
+Having multiple readers, even if only one thread can send commands, would be useful for visualization purposes.
+
+## Writing
+
+Let's create an instance of the `Robot`:
+
+```python
+import stretch_body.robot
+
+r = stretch_body.robot.Robot()
+assert r.startup()
+assert r.is_homed()
+```
+
+Then create a "writer", that will send motion commands:
+
+```python
+import random
+
+def write():
+    r.lift.move_to(random.uniform(0.3, 1.1))
+    r.arm.move_to(random.uniform(0.0, 0.4))
+    r.end_of_arm.move_to('wrist_roll', random.uniform(-1.0, 1.0))
+    r.head.move_to('head_pan', random.uniform(-1.0, 1.0))
+    r.push_command()
+```
+
+Then let's define a writing-only runner, that sends motion commands at 10hz:
+
+```python
+def wo_runner():
+    """write-only runner"""
+    while True:
+        write()
+        time.sleep(0.1)
+```
+
+### Results
+
+We can run `wo_runner()` in the main thread without any issues. But if we launch `wo_runner()` in a thread:
+
+```python
+import threading
+threading.Thread(target=wo_runner).start()
+```
+
+The program exits with no error. Likely an exception is being thrown in the thread, but isn't being surfaced to the user. Surfacing this exception is the first step towards making this library thread-safe.
+
+Notably, if we command out the lift and arm motion (leaving only the Dxl joints), the thread works fine.
+
```

Comment:
Again dynamixels don't run any of it's methods in async event loop threads. They just run on regular threads which is to be noted.

Reply:
Do you think asyncio could be the source of the silent error?

Reply:
May be. I think we should try your tests with the async param disabled to see what happens.